### PR TITLE
Merge latest blp_dev changes onto the WIP upstream merge

### PR DESCRIPTION
**Pull request by tylerparsons**
All features introduced by https://github.com/bloomberg/cartodb/pull/229

### CODE REVIEW

File: app/models/carto/user.rb [hunk 1]
```diff
@@ -222,15 +225,48 @@ def google_maps_client_id
   end
 
   # returns a list of basemaps enabled for the user
+  # when google map key is set it gets the basemaps inside the group "GMaps"
+  # if not it get everything else but GMaps in any case GMaps and other groups can work together
+  # this may have change in the future but in any case this method provides a way to abstract what
+  # basemaps are active for the user
+  #
```
Comment (acarrera42) [inline app/models/carto/user.rb:232]: Seems this comment is no longer valid if you remove the google_maps_enabled? logic below

File: app/models/carto/user.rb [hunk 2]
```diff
@@ -222,15 +225,48 @@ def google_maps_client_id
...
+  # Basemaps may also be filtered by profile attributes - this implementation respects such filters.
   def basemaps
-    (Cartodb.config[:basemaps] || []).select { |group| group != 'GMaps' || google_maps_enabled? }
```
Comment (acarrera42) [inline app/models/carto/user.rb:226]: Although we don't enable google maps this will lead to merge conflicts later.  Is there an issue with simply keeping google_maps_enabled to false?

**Review by acarrera42 — COMMENTED** [review]
Committing review comments.  Will continue reviewing remaining files later.

File: app/models/carto/user.rb (hunk 1 above)
Comment (tylerparsons) [reply app/models/carto/user.rb:232]: See the end of the function - google maps logic is kept but moved to the end.  I think this comment shows up as an addition to the merge branch because it may have been deleted in the upstream, but it's been in our repo for 3+ years so I kept it: https://github.com/bloomberg/cartodb/blame/blp_dev/app/models/carto/user.rb#L203

File: app/models/carto/user.rb (hunk 2 above)
Comment (tylerparsons) [reply app/models/carto/user.rb:226]: Logic kept and moved to the end of the function.

Comment (acarrera42) [reply app/models/carto/user.rb:226]: The logic appears to have moved to a new function filter_basemaps rather than the stay in the pre-existing base maps function.  Functionally it's moot since Bloomberg doesn't use google maps base layers.  I'm simply mentioning it for merging upstream carto code in the future.

File: app/models/user.rb [hunk 3]
```diff
@@ -1453,8 +1463,40 @@ def google_maps_client_id
...
 
   # returns a list of basemaps enabled for the user
+  # when google map key is set it gets the basemaps inside the group "GMaps"
+  # if not it get everything else but GMaps in any case GMaps and other groups can work together
+  # this may have change in the future but in any case this method provides a way to abstract what
+  # basemaps are active for the user
+  #
+  # Basemaps may also be filtered by profile attributes - this implementation respects such filters.
   def basemaps
-    (Cartodb.config[:basemaps] || []).select { |group| group != 'GMaps' || google_maps_enabled? }
```
Comment (acarrera42) [inline app/models/user.rb:1457]: Same comment as for model/user.rb

File: app/models/visualization/overlays.rb [hunk 4]
```diff
@@ -8,6 +8,8 @@ def initialize(visualization)
       end
 
       def create_default_overlays
+        prof_attrs = @visualization.user && @visualization.user.profile_attributes
+        puts "create_default_overlays - prof_attrs: #{prof_attrs}"
```
Comment (acarrera42) [inline app/models/visualization/overlays.rb]: Remove trace or place in log if useful

File: db/migrate/20170630090000_add_category_foreign_key.rb [hunk 5]
```diff
@@ -7,6 +7,6 @@
 
   down do
     execute "ALTER TABLE visualizations DROP CONSTRAINT visualizations_category_fkey;"
-    execute "ALTER TABLE visualizations DROP CONSTRAINT unique_id;"
+    execute "ALTER TABLE visualization_categories DROP CONSTRAINT unique_id;"
```
Comment (acarrera42) [inline db/migrate/20170630090000_add_category_foreign_key.rb:10]: Should migration scripts be modified?  Shouldn't we simply add a new migration script to make any changes?

File: db/migrate/20171114183148_profiles.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+Sequel.migration do
+
+  up do
```
Comment (acarrera42) [inline db/migrate/20171114183148_profiles.rb:3]: carto has a migration helper to avoid Postgres lockups: lib/carto/db/migration_helper.rb

Since these migrations were already applied though it may not make sense to modify these, but consider using the helper going forward.

File: db/migrate/20170630090000_add_category_foreign_key.rb (hunk 5 above)
Comment (mbektash) [reply db/migrate/20170630090000_add_category_foreign_key.rb:10]: this one is a bug fix actually

File: db/migrate/20171114183148_profiles.rb (hunk 6 above)
Comment (tylerparsons) [reply db/migrate/20171114183148_profiles.rb:3]: Good to know. I'll take a look in the future but in general the overhead of adding a column should be minimal

File: db/migrate/20170630090000_add_category_foreign_key.rb (hunk 5 above)
Comment (tylerparsons) [reply db/migrate/20170630090000_add_category_foreign_key.rb:10]: Correct, and it is also a change to the down task, not to the up task for an applied migration script, so it should be modified.

File: app/models/visualization/overlays.rb (hunk 4 above)
Comment (tylerparsons) [reply app/models/visualization/overlays.rb]: Will remove - didn't mean to include this.

File: app/models/carto/user.rb (hunk 2 above)
Comment (tylerparsons) [reply app/models/carto/user.rb:226]: `filter_basemaps` is a helper function within the scope of `basemaps` - the google maps logic still appears in the last line of `basemaps`.

File: db/migrate/20171114183148_profiles.rb (hunk 6 above)
Comment (gfiorav) [reply db/migrate/20171114183148_profiles.rb:3]: The helper won't interfere with normal operations. It should be used always for safety!

No need in your installation though, just looking for consistency.

File: app/controllers/api/json/profile_attributes_controller.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+
+class Api::Json::ProfileAttributesController < Api::ApplicationController
+  include CartoDB
+
+  def show
+    render json: current_user.nil? ? {} : current_user.profile_attributes
```
Comment (gfiorav) [inline app/controllers/api/json/profile_attributes_controller.rb]: Minor: you can rewrite as:

```ruby
render json: current_user.try(:profile_attributes) || {}
```

File: app/controllers/api/json/profile_attributes_controller.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+
+class Api::Json::ProfileAttributesController < Api::ApplicationController
```
Comment (gfiorav) [inline app/controllers/api/json/profile_attributes_controller.rb:2]: We use the `carto/` paths/modules for new controllers

File: app/controllers/api/json/profile_attributes_controller.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+  include CartoDB
```
Comment (gfiorav) [inline app/controllers/api/json/profile_attributes_controller.rb]: Why is this needed?

File: app/controllers/carto/api/visualizations_controller.rb [hunk 10]
```diff
@@ -355,12 +355,16 @@ def count
         type = params.fetch(:type, 'datasets')
         typeList = (type == 'datasets') ? "'table','remote'" : "'derived'"
         categoryType = (type == 'datasets') ? 1 : 2
+        samples_user_id = sample_maps_user.id if categoryType == 2 && sample_maps_user
```
Comment (gfiorav) [inline app/controllers/carto/api/visualizations_controller.rb:358]: Not strictly requesting a change, but adding a comment for readability:

It's not immediate to the reader that having a `samples_user_id` is tied to not only having `sample_maps_user` (which is the straightforward part) but to also `categoryType == 2` (not related to `sample_user` in any nature). Same happens above with the `samples` condition. 

With a software project this big and so many contributors, it's important to follow the "least surprise" principle, even at the expense of some extra lines :)

File: app/controllers/carto/api/visualizations_controller.rb [hunk 11]
```diff
@@ -355,12 +355,16 @@ def count
...
+        sample_type_counts = Object.new
```
Comment (gfiorav) [inline app/controllers/carto/api/visualizations_controller.rb:359]: Initializing a variable like this isn't to common in Ruby. If you must, you can initialize to `nil`, and then invoke it as

```ruby
sample_type_counts || default_value
```

and set it like
```ruby
sample_type_counts ||= "whatever overrite if not set"
```

File: app/controllers/carto/api/visualizations_controller.rb [hunk 12]
```diff
@@ -387,10 +391,29 @@ def count
...
+          query = "SELECT
+                COUNT(*) AS all,
+                SUM(CASE WHEN likes > 0 THEN 1 ELSE 0 END) AS liked
+              FROM (
+                SELECT results.*, (SELECT COUNT(*) FROM likes WHERE actor=? AND subject=results.id) AS likes FROM (
+                  SELECT v.id, v.type, v.category, v.locked
+                    FROM visualizations AS v
+                    WHERE v.user_id=? AND v.type IN (#{typeList}) AND v.locked=true
+                  ) AS results"
+          args = [user_id, user_id]
```
Comment (gfiorav) [inline app/controllers/carto/api/visualizations_controller.rb:406]: It's a good practice to initialize args right before using them.. so it's a bit more readable. Also, if they're not going to be used elsewhere, maybe you can save the extra variable.

```ruby
args = [user_id, user_id]
sample_type_counts = Sequel::Model.db.fetch(query, *args)
```

or 

```ruby
sample_type_counts = Sequel::Model.db.fetch(query, *[user_id, user_id])
```

File: app/controllers/carto/api/visualizations_controller.rb [hunk 13]
```diff
@@ -387,10 +391,29 @@ def count
...
+
+          query += ") AS results2"
```
Comment (gfiorav) [inline app/controllers/carto/api/visualizations_controller.rb:408]: Why can't you write that in the original query?

```ruby
         query = "SELECT
                COUNT(*) AS all,
                SUM(CASE WHEN likes > 0 THEN 1 ELSE 0 END) AS liked
              FROM (
                SELECT results.*, (SELECT COUNT(*) FROM likes WHERE actor=? AND subject=results.id) AS likes FROM (
                  SELECT v.id, v.type, v.category, v.locked
                    FROM visualizations AS v
                    WHERE v.user_id=? AND v.type IN (#{typeList}) AND v.locked=true
                  ) AS results ) AS resutls2"
```

File: app/models/carto/profile.rb [hunk 14]
```diff
@@ -0,0 +1,15 @@
+# encoding: UTF-8
+
+module Carto
+  class Profile < ActiveRecord::Base
+
+    validates :name, presence: true
+
+    has_many :profiles_users, :dependent => :destroy
```
Comment (gfiorav) [inline app/models/carto/profile.rb]: Rocket operators are mostly deprecated. You can use `Rubocop` to lint the file with our config and tell you about these things :)

Should read:

`has_many :profiles_users, dependent: :destroy`

File: app/models/carto/user.rb [hunk 15]
```diff
@@ -40,6 +40,9 @@ class Carto::User < ActiveRecord::Base
   has_many :layers_user
   has_many :layers, through: :layers_user, after_add: Proc.new { |user, layer| layer.set_default_order(user) }
 
+  has_many :profiles_user, dependent: :destroy, foreign_key: :user_id, inverse_of: :user
+  has_many :profiles, :through => :profiles_user
```
Comment (gfiorav) [inline app/models/carto/user.rb]: Same here:

```ruby
has_many :profiles, through: :profiles_users
```

File: db/migrate/20171114183148_profiles.rb [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+    create_table :profiles do
+      column :id, "uuid", :default=>Sequel::LiteralString.new("uuid_generate_v4()"),
+             :null=>false, :primary_key=>true
```
Comment (gfiorav) [inline db/migrate/20171114183148_profiles.rb:6]: Same rocket operartor thing here

File: db/migrate/20171114183148_profiles.rb [hunk 17]
```diff
@@ -0,0 +1,16 @@
+Sequel.migration do
+
+  up do
+    create_table :profiles do
+      column :id, "uuid", :default=>Sequel::LiteralString.new("uuid_generate_v4()"),
```
Comment (gfiorav) [inline db/migrate/20171114183148_profiles.rb:5]: Why not use the usual `uuid` generator?
 
```ruby
Uuid :id, primary_key: true, null: false, unique: false, default: 'uuid_generate_v4()'.lit
```

Looking at the latest migrations and following those conventions is best.

File: db/migrate/20171114183347_add_lock_flags_to_maps.rb [hunk 18]
```diff
@@ -0,0 +1,10 @@
+Sequel.migration do
+
+  change do
```
Comment (gfiorav) [inline db/migrate/20171114183347_add_lock_flags_to_maps.rb:3]: Write explicit `def up` and `def down` methods, or this behavior will be subject to whatever the next sequel gem update brings. Explicit is always best :)

File: app/models/visualization/overlays.rb (hunk 4 above)
Comment (gfiorav) [reply app/models/visualization/overlays.rb]: Rubocop will aid in preventing things like this too! Please check it out

**Review by gfiorav — CHANGES_REQUESTED** [review]
Hi, 

Next time please make sure you request a review on the minimum set of changes (your code vs the branch you based it off) prior to merging. It's also helpful to chop it up into several PRs. A PR this big (and with unrelated changes) is almost impossible to attain in one go, making it very hard to spot any issues with the code that might prevent bugs :)

I've left some comments but I'm also missing specs for the new models and controllers.

File: app/controllers/api/json/profile_attributes_controller.rb (hunk 9 above)
Comment (tylerparsons) [reply app/controllers/api/json/profile_attributes_controller.rb]: `include CartoDB` might not be needed.

File: db/migrate/20171114183148_profiles.rb (hunk 16 above)
Comment (tylerparsons) [reply db/migrate/20171114183148_profiles.rb:6]: I can change, but is it advisable to change an already applied migration script though?  Seems unnecessary to risk changing the behavior

File: db/migrate/20171114183347_add_lock_flags_to_maps.rb (hunk 18 above)
Comment (tylerparsons) [reply db/migrate/20171114183347_add_lock_flags_to_maps.rb:3]: I did make sure to test this beforehand with `db:migrate:up` and `db:migrate:down`.  If you think Sequel would introduce breaking changes to this functionality, I can change it.

File: db/migrate/20171114183148_profiles.rb (hunk 17 above)
Comment (tylerparsons) [reply db/migrate/20171114183148_profiles.rb:5]: Good to know in the future, but same concern as in `db/migrate/20171114183148_profiles.rb` - is it a good idea to change already applied migration scripts?

Comment (gfiorav) [reply db/migrate/20171114183148_profiles.rb:5]: Reviews should happen before the code is deployed :)

Agreed that it's late now

File: db/migrate/20171114183148_profiles.rb (hunk 16 above)
Comment (gfiorav) [reply db/migrate/20171114183148_profiles.rb:6]: Same, please say that this code is already in prod from the start (maybe you said it and I forgot). Not much point to CRing now.

File: db/migrate/20171114183347_add_lock_flags_to_maps.rb (hunk 18 above)
Comment (gfiorav) [reply db/migrate/20171114183347_add_lock_flags_to_maps.rb:3]: Explicit is always safer

File: db/migrate/20171114183148_profiles.rb (hunk 17 above)
Comment (tylerparsons) [reply db/migrate/20171114183148_profiles.rb:5]: We actually did review in 4 separate pulls on bloomberg/cartodb - I think you were just not added as a reviewer at the time.

Comment (gfiorav) [reply db/migrate/20171114183148_profiles.rb:5]: That's great! 

Then don't request my review on this massive PR please, or be explicit about *what* you want me to review. No point in me investing time in this if it's already reviewed. I'd love to be a reviewer, but you need to ping me before the code is merged in order for the review to be as helpful as possible.

**Comment by tylerparsons** [comment]
Thanks @gfiorav. I should have been clearer in my original PR description, sorry for any confusion and duplicated efforts.  All the features in this PR were reviewed separately on bloomberg/cartodb, however there was a good bit of refactoring in merging the production code so I wanted to get your opinion on whether these pieces still fit well into the merge banch. Should have elaborated on the context of https://github.com/bloomberg/cartodb/pull/229

**Comment by tylerparsons** [comment]
@gfiorav  I addressed most of those comments, agreed that these will be good improvements to the even though the code is deployed already.  Regarding specs, for better or worse, these changes were rolled without them because the tests for existing models/controllers were broken.  We have backlog items to fix these and write tests for this functionality as part of our next deployment.

**Comment by gfiorav** [comment]
Hey @tylerparsons,

Yes, don't worry! It was just a bit of miscommunication :)

I can't really tell the refactors from the final code since I'm not familiar with the original one you made. My guess is that nothing is broken, but please test it thoroughly.

Thanks

**Comment by tylerparsons** [comment]
Yes, manual testing has been quite thorough (painfully so) so we will definitely get those auto tests working :)
